dotnet/src/Connectors/Connectors.GoogleVertexAI/ToolCallBehavior.cs

```diff
@@ -0,0 +1,189 @@
+    /// </summary>
+    /// <remarks>
+    /// If no <see cref="Kernel"/> is available, no function information will be provided to the model.
+    /// </remarks>
+    public static ToolCallBehavior EnableKernelFunctions => new KernelFunctions(autoInvoke: false);
+
+    /// <summary>
+    /// Gets an instance that will both provide all of the <see cref="Kernel"/>'s plugins' function information
+    /// to the model and attempt to automatically handle any function call requests.
+    /// </summary>
+    /// <remarks>
+    /// When successful, tool call requests from the model become an implementation detail, with the service
+    /// handling invoking any requested functions and supplying the results back to the model.
+    /// If no <see cref="Kernel"/> is available, no function information will be provided to the model.
+    /// </remarks>
+    public static ToolCallBehavior AutoInvokeKernelFunctions => new KernelFunctions(autoInvoke: true);
+
+    /// <summary>Gets an instance that will provide the specified list of functions to the model.</summary>
+    /// <param name="functions">The functions that should be made available to the model.</param>
+    /// <param name="autoInvoke">true to attempt to automatically handle function call requests; otherwise, false.</param>
+    /// <returns>
+    /// The <see cref="ToolCallBehavior"/> that may be set into <see cref="ToolCallBehavior"/>
+    /// to indicate that the specified functions should be made available to the model.
+    /// </returns>
+    public static ToolCallBehavior EnableFunctions(IEnumerable<GeminiFunction> functions, bool autoInvoke = false)
+    {
+        Verify.NotNull(functions);
+        return new EnabledFunctions(functions, autoInvoke);
+    }
+
+    /// <summary>Initializes the instance; prevents external instantiation.</summary>
+    private ToolCallBehavior(bool autoInvoke)
+    {
+        this.MaximumAutoInvokeAttempts = autoInvoke ? DefaultMaximumAutoInvokeAttempts : 0;
+    }
+
+    /// <summary>Gets how many requests are part of a single interaction should include this tool in the request.</summary>
+    /// <remarks>
+    /// This should be greater than or equal to <see cref="MaximumAutoInvokeAttempts"/>. It defaults to <see cref="int.MaxValue"/>.
+    /// Once this limit is reached, the tools will no longer be included in subsequent retries as part of the operation, e.g.
+    /// if this is 1, the first request will include the tools, but the subsequent response sending back the tool's result
+    /// will not include the tools for further use.
+    /// </remarks>
+    public int MaximumUseAttempts { get; set; } = int.MaxValue;
+
+    /// <summary>Gets how many tool call request/response roundtrips are supported with auto-invocation.</summary>
+    /// <remarks>
+    /// To disable auto invocation, this can be set to 0.
+    /// </remarks>
+    public int MaximumAutoInvokeAttempts { get; set; }
```

Comment:
Because it's an implementation detail that still needs to be vetted, e.g. https://github.com/microsoft/semantic-kernel/pull/4686#discussion_r1464097022. Even as [Experimental], we should be surfacing only the surface area we think we're going to keep as stable.

Reply:
Okay, I will change these to private.

Reply:
done